draft-ietf-mls-protocol.md

```diff
@@ -1103,42 +1103,60 @@ opaque key_id<0..2^16-1>;
 
 ## Parent Hash
 
-The `parent_hash` extension serves to bind a KeyPackage to all the nodes
-above it in the group's ratchet tree. This enforces the tree invariant, meaning
-that malicious members can't lie about the state of the ratchet tree when they
-send Welcome messages to new members.
+The `parent_hash` extension serves to bind a KeyPackage inserted as part of an
+`UpdatePath` during a commit to the rest of the HPKE keys in that update path.
+Moreover, for each of those keys `parent_hash` also binds the set of parties
+that were told the corresponding secret key. This helps new members joining a group
+verify the tree invariant. That is, it prevents malicious members from
+constructing an artificial ratchet tree with a node v whose HPKE secret key is
+known to the malicious member yet where the member is not assigned a leaf in the
+subtree rooted at v. (Indeed, such a ratchet tree would violate the tree
+invariant.)
+
 
 ~~~~~
 opaque parent_hash<0..255>;
 ~~~~~
 
-This extension MUST be present in all Updates that are sent as part of a Commit
-message. If the extension is present, clients MUST verify that `parent_hash`
-matches the hash of the leaf's parent node when represented as a ParentNode
-struct.
+This extension MUST be present in the `leaf_key_package` Key Package field of an
+`UpdatePath` object. When processing a Commit message clients MUST recompute the 
+expected value of `parent_hash` for the committer's new leaf and verify that it 
+matches the `parent_hash` value in the `leaf_key_package`. Moreover, when joining
+a group new members MUST verify that if a leaf contains a `parent_hash` value than
+it matches the value obtained by recomputing `parent_hash` at the leaf.
 
-<!-- OPEN ISSUE: This scheme, in which the tree hash covers the parent hash, is
-designed to allow for more deniable deployments, since a signature by a member
-covers only its direct path. The other possible scheme, in which the parent hash
-covers the tree hash, provides better group agreement properties, since a
-member's signature covers the entire membership of the trees it is in. Further
-discussion is needed to determine whether the benefits to deniability justify
-the harm to group agreement properties, or whether there are alternative
-approaches to deniability that could be compatible with the other approach. -->
+The `parent_hash` at the root is `0`. To compute the parent hash at a non-root
```

Comment:
```suggestion
The `parent_hash` at the root is a zero-length bit string. To compute the parent hash at a non-root
```

Reply:
I've always been trying to use zero-length octet strings instead of "0" because "0" is ambiguous to me. For `Nh`-long all-zero string, I don't think we've used that anywhere else. Probably just because it's wordier

Reply:
My (weak) preference is to use an Nh long all-zero string. But only because that makes it the right data type for the output of a hash function (in contrast to the empty string).